src/transformers/models/pegasus_x/modeling_pegasus_x.py

```diff
@@ -0,0 +1,1659 @@
+
+        Returns:
+            output of shape `[batch_sizes, length, features]`. where length will be padded to a multiple of block_size
+        """
+        # [B, H, G+P, F]
+        global_and_local_k = torch.cat([global_k, local_k], dim=2)
+        # [B, H, G+P, F]
+        global_and_local_v = torch.cat([global_v, local_v], dim=2)
+
+        # [B, G+P]
+        extended_mask = nn.functional.pad(mask, pad=(dim.G, 0), value=0)
+
+        # [B, H, G, G+P]
+        attn_weights = torch.einsum("BHGF,BHXF->BHGX", global_q, global_and_local_k)
+        attn_weights = attn_weights + extended_mask[:, None, None, :]
+        attn_probs = nn.functional.softmax(attn_weights, dim=-1)
+        attn_probs = nn.functional.dropout(attn_probs, p=self.dropout, training=self.training)
+
+        # [B, H, G, F]
+        attn_output = torch.einsum("BHGX,BHXF->BHGF", attn_probs, global_and_local_v)
+        return attn_output, attn_probs
+
+    def compute_local_attention_representations(
+        self, global_k, global_v, local_q, local_k, local_v, mask, dim: DimensionInfo
+    ):
+        """Compute attention representations for local tokens.
+
+        Local tokens will attend to both global tokens as well as all other tokens within the same local block. Hence,
+        we need to tile and concatenate the global tokens to every local block
+
+        Args:
+            global_k (`torch.FloatTensor`) of shape [B, H, G, F]: key vectors from global tokens
+            global_v (`torch.FloatTensor`) of shape [B, H, G, F]: value vectors from global tokens
+            local_q (`torch.FloatTensor`) of shape [B, H, P, F]: query vectors from local tokens
+            local_k (`torch.FloatTensor`) of shape [B, H, P, F]: key vectors from local tokens
+            local_v (`torch.FloatTensor`) of shape [B, H, P, F]: value vectors from local tokens
+            mask (`torch.FloatTensor`) of shape [B, P]: attention mask
+            dim (DimensionInfo): DimensionInfo wrapper for dimensions
+
+        Returns:
+            output of shape `[batch_sizes, length, features]`. where length will be padded to a multiple of block_size
+        """
+        # [B, H, N, K, F]
+        blocked_local_q = local_q.view(dim.B, dim.H, dim.N, dim.K, dim.F)
+        # [B, H, N, K, F]
+        blocked_local_k = local_k.view(dim.B, dim.H, dim.N, dim.K, dim.F)
+        # [B, H, N, K, F]
+        blocked_local_v = local_v.view(dim.B, dim.H, dim.N, dim.K, dim.F)
+
+        # [B, N, G+K]
+        extended_mask = nn.functional.pad(mask.view(dim.B, dim.N, dim.K), pad=(dim.G, 0), value=0)
+
+        # [B, H, N, K, G]
+        blocked_local2global = torch.einsum("BHNKF,BHGF->BHNKG", blocked_local_q, global_k)
+        # [B, H, N, K, K]
+        blocked_local2local = torch.einsum("BHNKF,BHNXF->BHNKX", blocked_local_q, blocked_local_k)
+
+        # [B, H, N, K, G+K]
+        attn_weights = torch.cat([blocked_local2global, blocked_local2local], dim=4)
```

Comment:
(nit) -> think it's easier to understand "the last dimension" than the "4th dimension"
```suggestion
        attn_weights = torch.cat([blocked_local2global, blocked_local2local], dim=-1)
```